src/3d/qgs3dsceneexporter.cpp

```diff
@@ -0,0 +1,524 @@
+{
+  const Qgs3DMapSettings &settings = terrain->map3D();
+  QgsChunkNode *root = terrain->rootNode();
+
+  // just use LoD0 for now
+  int levelOfDetails = 0;
+  QVector<QgsChunkNode *> leafs;
+  leafs << root;
+  for ( int i = 0; i < levelOfDetails; ++i )
+  {
+    QVector<QgsChunkNode *> nodes = leafs;
+    leafs.clear();
+    for ( QgsChunkNode *node : nodes )
+    {
+      node->ensureAllChildrenExist();
+      QgsChunkNode *const *children = node->children();
+      for ( int i = 0; i < 4; ++i )
+      {
+        if ( children[i] != nullptr )
+        {
+          leafs.push_back( children[i] );
+        }
+      }
+    }
+  }
+
+  QgsTerrainGenerator *generator = settings.terrainGenerator();
+  QgsTerrainTileEntity *terrainTile = nullptr;
+  QgsTerrainTextureGenerator *textureGenerator = terrain->textureGenerator();
+  QSize oldResolution = textureGenerator->textureSize();
+  textureGenerator->setTextureSize( QSize( mTerrainTextureResolution, mTerrainTextureResolution ) );
+  switch ( generator->type() )
+  {
+    case QgsTerrainGenerator::Dem:
+      for ( QgsChunkNode *node : leafs )
+      {
+        terrainTile = getDemTerrainEntity( terrain, node );
+        this->parseDemTile( terrainTile, textureGenerator );
+      }
+      break;
+    case QgsTerrainGenerator::Flat:
+      for ( QgsChunkNode *node : leafs )
+      {
+        terrainTile = getFlatTerrainEntity( terrain, node );
+        this->parseFlatTile( terrainTile, textureGenerator );
+      }
+      break;
+    // TODO: implement other terrain types
+    case QgsTerrainGenerator::Mesh:
+      break;
+    case QgsTerrainGenerator::Online:
+      break;
+  }
+  textureGenerator->setTextureSize( oldResolution );
+}
+
+QgsTerrainTileEntity *Qgs3DSceneExporter::getFlatTerrainEntity( QgsTerrainEntity *terrain, QgsChunkNode *node )
+{
+  QgsFlatTerrainGenerator *generator = dynamic_cast<QgsFlatTerrainGenerator *>( terrain->map3D().terrainGenerator() );
+  QgsChunkLoader *loader = generator->createSynchronousChunkLoader( node );
+  FlatTerrainChunkLoader *flatTerrainLoader = qobject_cast<FlatTerrainChunkLoader *>( loader );
+  if ( flatTerrainLoader == nullptr ) return nullptr;
+  // the entity we created should be deallocated when we deallocate the scene exporter
+  Qt3DCore::QEntity *entity = flatTerrainLoader->createEntity( this );
+  QgsTerrainTileEntity *tileEntity = qobject_cast<QgsTerrainTileEntity *>( entity );
+  return tileEntity;
+}
+
+QgsTerrainTileEntity *Qgs3DSceneExporter::getDemTerrainEntity( QgsTerrainEntity *terrain, QgsChunkNode *node )
+{
+  QgsTerrainTileEntity *tileEntity = nullptr;
+  // Just create a new tile (we don't need to export exact level of details as in the scene)
+  // create the entity synchronously and then it will be deleted once our scene exporter instance is deallocated
+  const Qgs3DMapSettings &map = terrain->map3D();
+  QgsDemTerrainGenerator *generator = static_cast<QgsDemTerrainGenerator *>( map.terrainGenerator() );
```

Comment:
I think it would be safer to make your own copy of the terrain generator... imagine that during the export there may be loading of terrain tiles in a background thread - and we may modify the resolution of that generator while it is doing some work.

Reply:
I didn't think about that. I added ->clone()

Reply:
At first I carelessly added ->clone() but that actually doesn't work because the height map generator gets its terrain resolution parameter from map.terrainGenerator()
This is the code from QgsDemTerrainTileLoader constructor:
```
QgsDemTerrainGenerator *generator = static_cast<QgsDemTerrainGenerator *>( map.terrainGenerator() );
heightMapGenerator = generator->heightMapGenerator();
mSkirtHeight = generator->skirtHeight();
```